src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp

```diff
@@ -22,168 +23,198 @@ namespace ann /** Artificial Neural Network. */ {
 
 template<typename InputDataType, typename OutputDataType>
 BilinearInterpolation<InputDataType, OutputDataType>::
-BilinearInterpolation():
-    inRowSize(0),
-    inColSize(0),
-    outRowSize(0),
-    outColSize(0),
-    depth(0),
-    batchSize(0)
-{
+  BilinearInterpolation():
+  inRowSize(0),
+  inColSize(0),
+  outRowSize(0),
+  outColSize(0),
+  depth(0),
+  batchSize(0)
+  {
   // Nothing to do here.
-}
+  }
 
 template<typename InputDataType, typename OutputDataType>
 BilinearInterpolation<InputDataType, OutputDataType>::
-BilinearInterpolation(
+  BilinearInterpolation(
     const size_t inRowSize,
     const size_t inColSize,
     const size_t outRowSize,
     const size_t outColSize,
     const size_t depth):
-    inRowSize(inRowSize),
-    inColSize(inColSize),
-    outRowSize(outRowSize),
-    outColSize(outColSize),
-    depth(depth),
-    batchSize(0)
-{
-  // Nothing to do here.
-}
+  inRowSize(inRowSize),
+  inColSize(inColSize),
+  outRowSize(outRowSize),
+  outColSize(outColSize),
+  depth(depth),
+  batchSize(0)
+  {
+    // Let Po denote any output matrix point. This value at this
+    // point will be determined by four points in the padded input
+    // matrix. Let these be denoted by [P1, P2, P3, P4]. This will
+    // form a unit square. So if we know the bottom right corner then
+    // we can determine the other points.
+    // Po = P1*d1 + P2*d2 + P3*d3 + P4*d4 where di's is a constant.
+    // The di's and Pi's depend on the dimensions of input matrix
+    // and output matrix. So, we can precompute these points in the
+    // constructor and use them directly in the forward and backward
+    // functions. In bilinear interpolation we will pad the input
+    // matrix to make the calculations more simple.
+  	double scaleRow = (double) inRowSize / (double) outRowSize;
+    double scaleCol = (double) inColSize / (double) outColSize;
+    coeffs_pre = arma::cube(2, 2, outRowSize * outColSize);
+    index_pre = arma::cube(1, 2, outRowSize * outColSize);
+
+    for (size_t i = 0; i < outRowSize; i++)
```

Comment:
```suggestion
    for (size_t i = 0; i < outRowSize; ++i)
```

See https://github.com/mlpack/mlpack/pull/2464 for more details.